### PR TITLE
Amortize cost of out of order merges by performing on each tick

### PR DESCRIPTION
**Pull request by robskillington**

### CODE REVIEW

File: storage/series/buffer.go [hunk 1]
```diff
@@ -660,20 +688,78 @@ func (b *dbBufferBucket) discardMerged() discardMergedResult {
 		}
 	}
 
+	var lastWriteAt time.Time
 	iter := b.opts.MultiReaderIteratorPool().Get()
 	iter.Reset(readers)
+
+	defer func() {
+		iter.Close()
+		for _, stream := range readers {
```
Comment (prateek) [inline storage/series/buffer.go]: nice, good catch!

File: storage/series/buffer.go [hunk 2]
```diff
@@ -660,20 +688,78 @@ func (b *dbBufferBucket) discardMerged() discardMergedResult {
...
+		b.resetEncoders()
+		b.resetBootstrapped()
+		b.empty = true
+	}()
+
+	if b.hasJustSingleEncoder() {
+		// Already merged as a single encoder
+		encoder := b.encoders[0].encoder
+		newBlock := b.opts.DatabaseBlockOptions().DatabaseBlockPool().Get()
+		newBlock.Reset(b.start, encoder.Discard())
```
Comment (prateek) [inline storage/series/buffer.go]: Do we have to put the encoder back into the pool here?

File: storage/series/buffer.go [hunk 3]
```diff
@@ -660,20 +688,78 @@ func (b *dbBufferBucket) discardMerged() discardMergedResult {
...
+	result, err := b.merge()
+	if err != nil {
+		return discardMergedResult{}, err
+	}
+
+	merged := b.encoders[0].encoder
 
 	newBlock := b.opts.DatabaseBlockOptions().DatabaseBlockPool().Get()
-	newBlock.Reset(b.start, encoder.Discard())
+	newBlock.Reset(b.start, merged.Discard())
```
Comment (prateek) [inline storage/series/buffer.go]: same question as above for the lifecycle of the encoder here

File: storage/series/buffer.go (hunk 2 above)
Comment (robskillington) [reply storage/series/buffer.go]: `encoder.Discard()` will recycle the encoder, it calls `.Close()` after stealing the buffer =]

File: storage/series/buffer.go (hunk 3 above)
Comment (robskillington) [reply storage/series/buffer.go]: `.Discard()` recycles the encoder 👍

**Pull request merged**